### ISSUE TITLE
Rule no_unneeded_curly_braces does not understand annotation scoping

### ISSUE THREAD

**Issue by Chekote** [PHP-CS-Fixer/PHP-CS-Fixer#5899]
## Bug report

PHP-CS-Fixer's `no_unneeded_curly_braces` removes curly braces that are actually needed. Specifically, those that are defining the scope of annotations.

   * PHP version: 7.4.21
   * PHP CS Fixer version: 2.18.4
   * the command used to run PHP CS Fixer (run with `-vvv`): `php-cs-fixer my-file.php --config=.php_cs.dist --allow-risky=yes`
   * the configuration (file) you used:

```php
<?php

$finder = PhpCsFixer\Finder::create()->in(__DIR__);

return PhpCsFixer\Config::create()
    ->setRules([
        '@Symfony'               => true,
        'binary_operator_spaces' => [
            'operators' => [
                '=>' => 'align_single_space_minimal',
            ],
        ],
        'echo_tag_syntax'             => false,
        'ordered_class_elements'      => true,
        'ordered_imports'             => true,
        'yoda_style'                  => false,
        'blank_line_before_statement' => true,
        'class_attributes_separation' => [
            'elements' => [
                'const',
                'method',
                'property'
            ],
        ],
        'no_alternative_syntax'  => false,
        'phpdoc_no_empty_return' => true,
        'concat_space'           => [
            'spacing' => 'one'
        ],
        'phpdoc_no_alias_tag'    => [
            'type' => 'var',
            'link' => 'see',
        ],
        'single_line_throw'      => false
    ])
    ->setFinder($finder)->setUsingCache(false);
```

### Code snippet that reproduces the problem

```php
        /* @noinspection PhpFieldAssignmentTypeMismatchInspection */
        {
            $this->qa = $env->getContext(QualityAssurance::class);
            $this->store = $env->getContext(StoreContext::class);
            $this->web = $env->getContext(WebContext::class);
        }
```

Expected output:

```php
        /* @noinspection PhpFieldAssignmentTypeMismatchInspection */
        {
            $this->qa = $env->getContext(QualityAssurance::class);
            $this->store = $env->getContext(StoreContext::class);
            $this->web = $env->getContext(WebContext::class);
        }
```

Actual output:

```php
        /* @noinspection PhpFieldAssignmentTypeMismatchInspection */

        $this->qa = $env->getContext(QualityAssurance::class);
        $this->store = $env->getContext(StoreContext::class);
        $this->web = $env->getContext(WebContext::class);

```

### Repurcussions of the problem

This results in the `@noinspection` annotation only being applied to the first line:

Before:

![Screen Shot 2021-08-26 at 7 51 54 AM](https://user-images.githubusercontent.com/550486/130965897-c995569a-27e5-49ba-8067-ba8ec9adaba7.png)

After:

![Screen Shot 2021-08-26 at 7 53 01 AM](https://user-images.githubusercontent.com/550486/130966007-3bc88b02-c55b-4459-900f-1b6a1ff46854.png)

**Comment by kubawerlos** [comment]
The description of the fixer is:

> Removes unneeded curly braces that are superfluous and aren't part of a control structure's body.

And in the examples above it does exactly what it suppose to.

**Comment by Chekote** [comment]
@kubawerlos sure, if you want to take a dogmatic approach to that description and enforce it word for word, then yes you're right.

But do you think the scenario I described in my bug is the explicit intent of the fixer, or simply an oversight? I would argue that even though the wording of the fixer's description does not cover this scenario, if you were to examine the spirit of the fixer rather than the letter, it should.

I believe that this is simply a scenario that the author did not consider, and therefore did not cover in their description. I do not believe it was an explicit intent to remove curly braces in this scenario.

**Comment by kubawerlos** [comment]
The scenario for me is the classic "I made something bad and it went bad".

I'd focus on solving the root problem - having `@noinspection` in the first place.

If I had `@noinspection` in my code and problem with `no_unneeded_curly_braces` fixer I would worry about the first one, not the latter one.

**Comment by Chekote** [comment]
Firstly, there's no need to be rude. You do not know the full story behind the code, so jumping to the conclusion that I made something bad is quite inappropriate. I'm not going to respond to that comment any further since it's beside the point.

Secondly, you are missing the point. It doesn't matter what the annotation is. My bug report is regarding the fact that when you use curly braces to scope an annotation to multiple lines, then PHP-CS-Fixer removes it even though it has functionality.

**Comment by kubawerlos** [comment]
Saying that using `@noinspection` is bad is not being rude, it's just stating the fact. Also, I did not wrote you did this anywhere...

Can you elaborate a bit more about these annotation scoping? I've never met this practice in any of dozens project I was participating in.

I'd still stay on claiming that the fixer here is not a problem, but tricking some other tool (phpcs?) with `@noinspection` - it's like having a car with broken brakes and stain on the seat and worrying about the stain.

**Comment by keradus** [comment]
other question - why `/* @foo */` and not `/** @foo */` ? shouldn't PHPDoc annotation be part of PHPDoc comment and not regular comment?
does the problem appears for both types of comments?

**Comment by Chekote** [comment]
@kubawerlos this is a feature used in PHPStorm. I believe the annotations are respected by other tools also, but when you encounter a warning in PHPStorm such as shown below:

![Screen Shot 2021-08-26 at 7 53 01 AM](https://user-images.githubusercontent.com/550486/130966007-3bc88b02-c55b-4459-900f-1b6a1ff46854.png)

You can bring up the context menu for various ways to deal with it. A few of the options are to suppress it for the statement, method, class, etc. If you choose for statement, then the annotation will appear immediately above the statement.

If you have multiple statements in a row, as in the example above. You can either place an annotation individually above each statement or use curly braces to scope the annotation to more than one statement.

@keradus both work, and PHPStorm actually uses `/** @foo */` when automatically adding the annotation, but funnily enough php-cs-fixer replaces `/** @foo */` with `/* @foo */`. I assume it is because php-cs-fixer does not like docblocks on statements.

**Comment by Chekote** [comment]
@kubawerlos in regards to your comment:

> Saying that using @noinspection is bad is not being rude, it's just stating the fact. Also, I did not wrote you did this anywhere...

I would really prefer to focus on the discussion at hand, but since you insist on bringing it up again:

I disagree. If PHP had Generics, then yes, I would agree that using `@noinspection` is objectively bad. But without Generics, it is at worst a code smell. The complexity developers need to add to a codebase to work around the lack of Generics without using `@noinspection` offers little to no benefit.

For example, my code above is basically using an IoC container. Much like Symfony and Laravel projects would use. How do you suggest we deal with the fact that the IoC method will return `mixed` according to the type system, but we have 100% certainty that it will actually return a type matching the specified argument?

If I have to do a type check everywhere I use the IoC container, then that opens a whole can of worms. Firstly, the type check is wasteful complexity. Secondly, what do we do if the type check fails (which it never will)? Do we throw an exception? If so, now we have leaked that unnecessary complexity to all the callers because they now have to deal with that exception that will never actually occur. Or do we throw a PHP Error instead? If so, why? What's the point?

**Comment by Chekote** [comment]
@kubawerlos I have reached out to JetBrains support to get clarification on whether or not scoping annotations via curly braces is a supported feature.

**Comment by kubawerlos** [comment]
Once again - I'd focus on a root problem, which is using the container directly:

[> As easy as this seams to be, using the container directly is not considered a good practice because it hides the dependencies of your classes, making them coupled to external configuration, thus harder to test, harder to review, etc.](https://symfony.com/blog/new-in-symfony-3-4-services-are-private-by-default)

I know it's easy to say "don't do it" when apparently you are in situation it is already done, if I was there I'd plan to code how to get rid of that tech debt.

Maybe using [`assert`](https://www.php.net/manual/en/function.assert.php) is a workaround worth to try?

**Comment by Chekote** [comment]
What about situations such as how Eloquent delegates to other classes?

Even using a simple `where()` call on an Eloquent `Model` results in PHPStorm complaining that the method does not exist, because it is not actually defined on the Model class, but delegated to the `Builder` class via the magic `__call` method:

![Screen Shot 2021-08-30 at 12 49 53 PM](https://user-images.githubusercontent.com/550486/131382454-85bd2718-4788-4fee-8166-085c95f37aea.png)

So, I could use `Lab::query()->where()` here, but that only kicks the can further down the road, because there are plenty of other scenarios where you can get tangled on Eloquent's web and not have a way to make the type system happy without significantly complicating your query by separating it into variables with `@var` annotations.

So instead, to keep the code simple and clean in the spirit of Laravel, we can use the same scoping here:

![Screen Shot 2021-08-30 at 12 52 35 PM](https://user-images.githubusercontent.com/550486/131382798-dd83fdcd-2ef7-4fe6-9b63-1d35b2582b6e.png)

In the interests of completion, let's continue the experiment and try to use `query()`:

![Screen Shot 2021-08-30 at 12 56 43 PM](https://user-images.githubusercontent.com/550486/131383324-e48f3a3d-440b-409f-9e69-f6b2e74c306c.png)

PHPStorm is happy now because I am coincidentally only using methods that exist on the `Builder` returned by `query()`. But if the `Lab` uses the `SoftDeletes` trait, I also have access to the `withTrashed` method. So let's try to use that:

![Screen Shot 2021-08-30 at 12 58 13 PM](https://user-images.githubusercontent.com/550486/131383497-95d489bc-74f9-4fc0-b060-4c63b5a332f5.png)

Well crap, now PHPStorm is unhappy again. And this is particularly challenging because `Builder` does not have a `withTrashed` method, but it happily proxies the call to the `Lab` class which does have the method:

![Screen Shot 2021-08-30 at 1 03 20 PM](https://user-images.githubusercontent.com/550486/131384092-276b694d-6a15-4ffe-aa6c-5a0e8bda73d8.png)

So now we have two further problems:

* We are lying about the type of `$labQuery` because it is a `Builder`, not a `Lab`
* We still can't make PHPStorm happy, because `withTrashed` is declared as static

I could use `/** @var Builder|Lab $labQuery */`, but that's still wrong because `Builder` will only proxy called to query scopes, not all methods on the `Lab` class.

TL:DR: yes, you are right that some of the things I have to do are not ideal. But I am working within the constraints of the tools and frameworks that I have no option but to work with. So I am trying to do the best I can with it. I need `@noinspection` to allow me to remove some of the "noise" from the warnings so that developers can have faith that the warnings they see are legitimate. If I do not fix these, then they will lose faith in the warnings, and legitimate problems will get overlooked.

**Comment by SpacePossum** [comment]
I never knew this was a thing, so I didn't consider it when making the rule. If the PHPDoc (annotation) followed by a curly block for scoping is something people really use I see no reason not to support it. It will likely harder to stop the other rule from turning the PHPDocs into comments, than stopping this rule from removing the braces. You would need both as I would not support scoped comments, only PHPDocs, but that is just my preference.

**Comment by Chekote** [comment]
@SpacePossum that's fair. One problem at a time I suppose 😄

**Comment by Wirone** [comment]
Personally I would close this as it seems like upstream issue leaked as Fixer issue. For me mixing comment (which `@noinspection` actually is` with the code (additional curly braces) is bad. IDE should support block annotations like `@noinspection-start` / `@noinspection-end`, not requiring workarounds in the code to make it work. I have never seen such usage in my life and I would not approve it on code review 😉. It would be better to extract these lines to private method and put annotation there, to disable inspections for entire block of code.

@Chekote please let us know if you still need it, because core team's capacity is limited and I would rather like to close it to clean the backlog from issues with limited interest and niche problem to fix 🙂.

**Comment by vami7ir** [comment]
I ran into a similar problem, despite the fact that the documentation says that no_unneeded_curly_braces takes an array in the form ['namespaces' => true], instead it specified 'no_unneeded_curly_braces' => FALSE, after which curly braces were not removed.
![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/61095786/9a3dbde7-4174-42c5-823f-974a88807fca)
![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/61095786/35d063cf-02cb-42e1-9d6f-61f348088826)

**Comment by Wirone** [comment]
@vami7ir with `'no_unneeded_curly_braces' => FALSE` fixer is disabled, so it's expected behavior that it did not do anything. Also, with `['namespaces' => true]` it wouldn't fix your code too, because these curly braces are not related to namespace declaration. These curly braces are useless, but such case is not covered by this fixer.

**Comment by vami7ir** [comment]
But wasn't the question about removing curly braces unrelated to logical operators, then in my opinion this turns out to be a solution for preserving curly braces

**Comment by Wirone** [comment]
Closing due to lack of response to [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/5899#issuecomment-1542902972). We can always re-open if needed, but at this moment it looks like the problem lays somewhere else (and we don't have resources to work on such issues).
